src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs

```diff
@@ -35,81 +39,107 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
+                }
+
+                // Temporarily filter out suppress and configure code actions, as we'll later combine them under a top-level
+                // code action.
+                if (codeFix.Action is AbstractConfigurationActionWithNestedActions)
+                {
+                    continue;
+                }
+
+                results.Add(GenerateVSCodeAction(request, codeFix.Action, CodeActionKind.QuickFix));
             }
 
-            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            foreach (var (action, _) in codeRefactorings.SelectMany(codeRefactoring => codeRefactoring.CodeActions))
+            {
+                // Filter out code actions with options since they'll show dialogs and we can't remote the UI and the options.
+                if (action is CodeActionWithOptions)
+                {
+                    continue;
+                }
 
-            var textSpan = ProtocolConversions.RangeToTextSpan(selection, text);
-            var codeFixCollections = await codeFixService.GetFixesAsync(document, textSpan, true, cancellationToken).ConfigureAwait(false);
-            var codeRefactorings = await codeRefactoringService.GetRefactoringsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+                results.Add(GenerateVSCodeAction(request, action, CodeActionKind.Refactor));
+            }
 
-            var codeActions = codeFixCollections.SelectMany(c => c.Fixes.Select(f => f.Action)).Concat(
-                                codeRefactorings.SelectMany(r => r.CodeActions.Select(ca => ca.action)));
+            // Special case (also dealt with specially in local Roslyn): 
+            // If we have configure/suppress code actions, combine them under one top-level code action.
+            var configureSuppressActions = codeFixes.Where(a => a.Action is AbstractConfigurationActionWithNestedActions);
+            if (configureSuppressActions.Any())
+            {
+                results.Add(GenerateVSCodeAction(request, new CodeActionWithNestedActions(
+                    CodeFixesResources.Suppress_or_Configure_issues,
+                    configureSuppressActions.Select(a => a.Action).ToImmutableArray(), true), CodeActionKind.QuickFix));
+            }
 
-            // Flatten out the nested codeactions.
-            var nestedCodeActions = codeActions.Where(c => c is CodeAction.CodeActionWithNestedActions nc && nc.IsInlinable).SelectMany(nc => nc.NestedCodeActions);
-            codeActions = codeActions.Where(c => !(c is CodeAction.CodeActionWithNestedActions)).Concat(nestedCodeActions);
+            return results.ToArray();
 
-            return codeActions;
+            // Local functions
+            static VSCodeAction GenerateVSCodeAction(
+                CodeActionParams request,
+                CodeAction codeAction,
+                CodeActionKind codeActionKind,
+                string parentTitle = "")
```

Comment:
Hmm, tried this out per your suggestion - it seems to be a bit problematic since using a string builder makes handling nested code actions difficult due to a combination of recursion + using the same StringBuilder, i.e. all the nested options are concatenated with each other.